### DISCUSSION TITLE
Trouble Including Static Catchment Attributes

### DISCUSSION THREAD

**Discussion by hydro-coop** [neuralhydrology/neuralhydrology#207] [Q&A] [answered]
Hello. I was hoping to receive some help getting an LSTM running that includes static catchment attributes. I followed the guidelines of creating an attributes folder in my data_dir with unique csv files for each basin, and the files are indexed by basin_id (see image). However, when I list the attributes in my .yml file, I continue to receive an error message "The following attributes have a std of zero or NaN, which results in NaN's when normalizing the features. Remove the attributes from the attribute feature list and restart the run." It tells me I must remove every one of my static attributes. I have also included an image of the error message. My understanding is that they are static attributes and won't change over time. How can I get around this error?

Can anyone provide some guidance on how I can proceed? Thank you in advance.

![Screenshot 2024-11-22 094509](https://github.com/user-attachments/assets/a26847de-cff9-4821-bda9-9876c905fc86)
![Screenshot 2024-11-22 094446](https://github.com/user-attachments/assets/a684e858-390a-4b5f-b4ba-89ca44174f90)

**Comment by gauchm** [comment, marked as answer]
> with unique csv files for each basin

I don't really understand what that means. Why do you have separate csv files for each basin? From your screenshot, it looks like one file already contains the attributes for multiple basins.

My first guess for why you get a std of zero would be that you're trying to use static attributes when you're training on a single basin. Is that what you are doing (if yes, that's typically not ideal in the first place, see [this paper](https://hess.copernicus.org/articles/28/4187/2024/hess-28-4187-2024.html)).

If that's not the problem, then it'd be helpful if you could share your config.yml.

**Reply by hydro-coop** [reply]
I apologize for the confusing image. The screenshot is of one file that contains multiple basins, but I also have one basin per file saved. 

I think the second part of your reply is getting to the main issue I'm having. I was training on a single basin to ensure I can get everything to run properly, but I will add more basins for my work. I just ran some tests and it looks like if you includ static attributes for training a single basin, it will generate the error message. However, as soon as you train with more than one basin the error does not appear and it runs smoothly.

Thank you for your reply. Learning something for the first time is always an adventure.

**Reply by gauchm** [reply]
Great, glad we figured it out. For dev settings, I'd recommend keeping a "dev set" of a few basins to test things before starting the actual big runs. This way, the static arguments have a std > 0 (if the dev set is diverse enough). Good luck with your experiments :)

**Reply by hydro-coop** [reply]
Hi again, @gauchm . I was hoping to get some help regarding a similar topic in NH. I have been able to seamlessly implement static attributes, but now I am hoping to utilize the 'evolving_attributes' feature of NH to include some basin attributes that change over time. I have referenced the page of configuration arguments, and I believe I have everything structured correctly. However, I am still receiving an error message "cannot convert DataFrame with non-unique columns". I have double checked, I do not have repeated column names between 'static_attributes', 'dynamic_inputs', and 'evolving_attributes'. I will include screenshots below for better understanding. 

1. Config file reference to evolving attributes.
![ref_structure](https://github.com/user-attachments/assets/aba36266-533c-437c-96fd-1cf95e887941)

2. Structure of 'evolving_attributes.pkl' referenced in "additional_feature_files" (not all attributes shown)
![evolv_atts_struct](https://github.com/user-attachments/assets/3692f27e-9788-4f81-b39d-7bf80d76ccba)

3. Error message received when trying to train a model with cfg that includes "additional_feature_files" and "evolving_attributes"
![error_mssg](https://github.com/user-attachments/assets/4e6b7a8b-c414-4c5d-a00a-d5b3bac93d94)

**Reply by hydro-coop** [reply]
UPDATE: Error has been resolved. It seems to be relatively unrelated to the message "cannot convert DataFrame with non-unique columns". It turns out there was a single basin_id in my list of training basins that was not included in my evolving attributes file. Otherwise, following the description of how to setup 'evolving_attributes' and 'additional_feature_files' appears to be working.

**Reply by kratzert** [reply]
Thanks for the update. Seems like we could add a check here that results in a more verbose error. I think that would be relatively easy to add so if you are looking for an easy way to contribute, this could be a nice small PR :)

P.S.: Also looking forward to hearing from you about your results.

**Reply by hydro-coop** [reply]
Excited to share with you! We are getting close to having our final model results, so I will reach out then!
